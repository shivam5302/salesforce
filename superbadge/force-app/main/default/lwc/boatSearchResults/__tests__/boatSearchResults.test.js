import { createElement } from 'lwc';
import BoatSearchResults from 'c/boatSearchResults';
import getBoats from '@salesforce/apex/BoatDataService.getBoats';
import getBoatsMock from './data/getBoats.json';
import updateBoatList from '@salesforce/apex/BoatDataService.updateBoatList';
import { publish, MessageContext } from 'lightning/messageService';
import BOATMC from '@salesforce/messageChannel/BoatMessageChannel__c';

jest.mock('@salesforce/apex/BoatDataService.getBoats',
	() => {
		const { createApexTestWireAdapter } = require('@salesforce/sfdx-lwc-jest');
		return {
			default: createApexTestWireAdapter(jest.fn())
		};
	},
	{ virtual: true }
)

jest.mock('@salesforce/apex/BoatDataService.updateBoatList', () => {
	return {
		default: jest.fn()
	};
}, { virtual: true });

jest.mock('@salesforce/apex', () => ({ refreshApex: jest.fn() }), {
	virtual: true,
});

jest.mock('c/boatTile');
jest.mock('c/boatsNearMe');

const flushPromises = async () => new Promise((resolve) => setTimeout(resolve, 0));

describe('c-boat-search-results', () => {
	let element;
	beforeEach(()=>{
		element = createElement('c-boat-search-results', {
			is: BoatSearchResults
		});

		// Act
		document.body.appendChild(element);

		getBoats.emit(getBoatsMock);
	})
	afterEach(() => {
		// The jsdom instance is shared across test cases in a single file so reset the DOM
		while (document.body.firstChild) {
			document.body.removeChild(document.body.firstChild);
		}
		jest.clearAllMocks();
	});

	it('Should render the boats When gallery tab is active', async() => {
		//Given
		const gallelryTab = element.shadowRoot.querySelector('.gallery');
		const boatTile = element.shadowRoot.querySelector('c-boat-tile')
		

		//When
		gallelryTab.click();

		//Then
		await flushPromises();
		expect(boatTile.boat).toEqual(getBoatsMock[0]);
		expect(element).toMatchSnapshot();
	});

	it('Should render the boats When editor tab is active', async() => {
		//Given
		const editorTab = element.shadowRoot.querySelector('.editor');
		const lightningTable = element.shadowRoot.querySelector('lightning-datatable')
		
		//When
		editorTab.click();
		
		//Then
		await flushPromises();
		expect(lightningTable.data).toEqual(getBoatsMock);
	});

	it('Should render the boats When Near Me tab is active', async() => {
		//Given
		const nearmeTab = element.shadowRoot.querySelector('.nearme');
		const boatsNearMe = element.shadowRoot.querySelector('c-boats-near-me')
		element.searchBoats(getBoatsMock[0].BoatType__c)
		
		//When
		nearmeTab.click();
		
		//Then
		await flushPromises();
		expect(boatsNearMe.boatTypeId).toBe(getBoatsMock[0].BoatType__c);
	});

	it('Should show toast error or wire method', async() => {
		//Given
		const mockError = { body: { message: 'Some error' }, status: 500 };
		const toastHandler = jest.fn();
		element.addEventListener('lightning__showtoast', toastHandler);
		
		//When
		getBoats.error(mockError);
		
		//Then
		await flushPromises();
		expect(toastHandler).toHaveBeenCalledTimes(1);

		const toastEvent = toastHandler.mock.calls[0][0];
		expect(toastEvent.detail).toEqual({
			title: 'Error',
			message: 'Getting error while fetching boats',
			variant: 'error'
		});
	});



	it('should dispatch success toast when updateBoatList succeeds', async () => {
		// Mock Apex success
		updateBoatList.mockResolvedValue({});

		// Add toast listener
		const toastHandler = jest.fn();
		element.addEventListener('lightning__showtoast', toastHandler);

		// Simulate save event
		const datatable = element.shadowRoot.querySelector('lightning-datatable');
		const saveEvent = new CustomEvent('save', {
			detail: { draftValues: getBoatsMock[0] },
			bubbles: true,
			composed: true
		});
		datatable.dispatchEvent(saveEvent);

		await flushPromises();

		// Assertions
		expect(toastHandler).toHaveBeenCalledTimes(1);

		const toastEvent = toastHandler.mock.calls[0][0];
		expect(toastEvent.detail).toEqual({
			title: 'Success',
			message: 'Data saved successfully',
			variant: 'success'
		});
	});


	it('should dispatch err toast when updateBoatList failed', async () => {
		// Mock Apex success
		const mockError = 'Something went wrong';
		updateBoatList.mockRejectedValue(mockError);


		// Add toast listener
		const toastHandler = jest.fn();
		element.addEventListener('lightning__showtoast', toastHandler);

		// Simulate save event
		const datatable = element.shadowRoot.querySelector('lightning-datatable');
		const saveEvent = new CustomEvent('save', {
			detail: { draftValues: getBoatsMock[0] },
			bubbles: true,
			composed: true
		});
		datatable.dispatchEvent(saveEvent);

		await flushPromises();

		// Assertions
		expect(toastHandler).toHaveBeenCalledTimes(1);

		const toastEvent = toastHandler.mock.calls[0][0];
		expect(toastEvent.detail).toEqual({
			title: 'Error',
			message: 'Error while saving data '+mockError,
			variant: 'error'
		});
	});


	it('should call publish with correct payload and message channel', async() => {

		const context = 'context';
		MessageContext.emit(context);
		const boatTile = element.shadowRoot.querySelector('c-boat-tile');
		const event = new CustomEvent('boatselect', {
			detail: {
				boatId: 'testId'
			},
			bubbles: true,
			composed: true
		});
		
		// ðŸ”¥ Dispatch event
		boatTile.dispatchEvent(event);
		
		await flushPromises();
		console.log(publish.mock.calls[0]);
		
		// âœ… Assert publish was called correctly
		expect(publish).toHaveBeenCalledWith(
			context,
			BOATMC,
			{ recordId: 'testId' }
		);
	});
});