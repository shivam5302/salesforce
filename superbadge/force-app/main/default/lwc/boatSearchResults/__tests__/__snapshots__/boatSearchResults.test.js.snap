// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-boat-search-results Should render the boats When gallery tab is active 1`] = `
<c-boat-search-results>
  #shadow-root(open)
    <lightning-card>
      #shadow-root(open)
        <slot />
        <slot
          name="actions"
        />
        <slot
          name="footer"
        />
        <slot
          name="title"
        />
      <lightning-tabset>
        #shadow-root(open)
          <slot />
        <lightning-tab
          class="gallery"
        >
          #shadow-root(open)
            <slot />
          <div
            class="slds-scrollable_y"
          >
            <lightning-layout
              class="slds-wrap"
            >
              #shadow-root(open)
                <slot />
              <lightning-layout-item
                class="slds-p-around_small slds-size_1-of-4"
              >
                #shadow-root(open)
                  <slot />
                <c-boat-tile>
                  #shadow-root(open)
                </c-boat-tile>
              </lightning-layout-item>
            </lightning-layout>
          </div>
        </lightning-tab>
        <lightning-tab
          class="editor"
        >
          #shadow-root(open)
            <slot />
          <lightning-datatable>
            #shadow-root(open)
          </lightning-datatable>
        </lightning-tab>
        <lightning-tab
          class="nearme"
        >
          #shadow-root(open)
            <slot />
          <c-boats-near-me>
            #shadow-root(open)
          </c-boats-near-me>
        </lightning-tab>
      </lightning-tabset>
    </lightning-card>
</c-boat-search-results>
`;
